lib/view_component/base.rb

```diff
@@ -218,9 +218,15 @@ def format
 
     # Use the provided variant instead of the one determined by the current request.
     #
+    # @deprecated Define separate per-variant templates instead. Will be removed in v3.0.0.
     # @param variant [Symbol] The variant to be used by the component.
     # @return [self]
     def with_variant(variant)
+      ActiveSupport::Deprecation.warn(
+        "`with_variant` is deprecated and will be removed in ViewComponent v3.0.0.\n\n" \
+        "Define separate per-variant templates instead."
```

Comment:
I'm not sure if this is the guidance we want to provide. Perhaps we could leave this line out for now and see if folks ask for guidance on alternatives?

Reply:
Ok sounds good! 😄

Reply:
what is the alternative for this?

Reply:
@Sohair63 I think it depends on your specific use-case, but the general guidance is:

* https://guides.rubyonrails.org/layouts_and_rendering.html#the-variants-option is what powers the variant functionality in view component. (The only change we made is disabling explicitly setting variants on a component.)
* You can pass an argument to the component so it renders different content.
* You can conditionally render a different component based on some condition in your code.